File: stacks/python/base/code/errors.tpl.py
"""
File: errors.tpl.py
Purpose: Structured error handling for FastAPI applications
Generated for: {{PROJECT_NAME}}
"""

from typing import Any, Dict, Optional
from fastapi import HTTPException, Request
from fastapi.responses import JSONResponse
from pydantic import BaseModel
import logging

logger = logging.getLogger(__name__)


class ErrorResponse(BaseModel):
    error: str
    message: str
    details: Optional[Dict[str, Any]] = None
    request_id: Optional[str] = None


class AppError(Exception):
    """Base application error"""
    def __init__(
        self,
        message: str,
        error_code: str = "INTERNAL_ERROR",
        status_code: int = 500,
        details: Optional[Dict[str, Any]] = None,
    ):
        self.message = message
        self.error_code = error_code
        self.status_code = status_code
        self.details = details
        super().__init__(message)


class NotFoundError(AppError):
    def __init__(self, resource: str, resource_id: str):
        super().__init__(
            message=f"{resource} with id '{resource_id}' not found",
            error_code="NOT_FOUND",
            status_code=404,
            details={"resource": resource, "id": resource_id},
        )


class ValidationError(AppError):
    def __init__(self, message: str, details: Optional[Dict[str, Any]] = None):
        super().__init__(
            message=message,
            error_code="VALIDATION_ERROR",
            status_code=400,
            details=details,
        )


class AuthenticationError(AppError):
    def __init__(self, message: str = "Authentication required"):
        super().__init__(
            message=message,
            error_code="UNAUTHORIZED",
            status_code=401,
        )


class AuthorizationError(AppError):
    def __init__(self, message: str = "Permission denied"):
        super().__init__(
            message=message,
            error_code="FORBIDDEN",
            status_code=403,
        )


class RateLimitError(AppError):
    def __init__(self, retry_after: int = 60):
        super().__init__(
            message="Rate limit exceeded",
            error_code="RATE_LIMITED",
            status_code=429,
            details={"retry_after": retry_after},
        )


# FastAPI exception handlers
async def app_error_handler(request: Request, exc: AppError) -> JSONResponse:
    request_id = getattr(request.state, "request_id", None)
    
    logger.error(
        f"AppError: {exc.error_code} - {exc.message}",
        extra={"request_id": request_id, "details": exc.details},
    )
    
    return JSONResponse(
        status_code=exc.status_code,
        content=ErrorResponse(
            error=exc.error_code,
            message=exc.message,
            details=exc.details,
            request_id=request_id,
        ).model_dump(),
    )


async def unhandled_error_handler(request: Request, exc: Exception) -> JSONResponse:
    request_id = getattr(request.state, "request_id", None)
    
    logger.exception(
        f"Unhandled error: {str(exc)}",
        extra={"request_id": request_id},
    )
    
    return JSONResponse(
        status_code=500,
        content=ErrorResponse(
            error="INTERNAL_ERROR",
            message="An unexpected error occurred",
            request_id=request_id,
        ).model_dump(),
    )


def register_error_handlers(app):
    """Register error handlers with FastAPI app"""
    app.add_exception_handler(AppError, app_error_handler)
    app.add_exception_handler(Exception, unhandled_error_handler)

File: _complete_archive/_archive_2026-02-26_18-50-32/stacks/node/base/code/errors.tpl.js
/*
File: errors.tpl.js
Purpose: Structured error handling for Express/Node.js applications
Generated for: {{PROJECT_NAME}}
*/

class AppError extends Error {
    constructor(message, errorCode = 'INTERNAL_ERROR', statusCode = 500, details = null) {
        super(message);
        this.name = 'AppError';
        this.errorCode = errorCode;
        this.statusCode = statusCode;
        this.details = details;
        Error.captureStackTrace(this, this.constructor);
    }
}

class NotFoundError extends AppError {
    constructor(resource, resourceId) {
        super(
            `${resource} with id '${resourceId}' not found`,
            'NOT_FOUND',
            404,
            { resource, id: resourceId }
        );
        this.name = 'NotFoundError';
    }
}

class ValidationError extends AppError {
    constructor(message, details = null) {
        super(message, 'VALIDATION_ERROR', 400, details);
        this.name = 'ValidationError';
    }
}

class AuthenticationError extends AppError {
    constructor(message = 'Authentication required') {
        super(message, 'UNAUTHORIZED', 401);
        this.name = 'AuthenticationError';
    }
}

class AuthorizationError extends AppError {
    constructor(message = 'Permission denied') {
        super(message, 'FORBIDDEN', 403);
        this.name = 'AuthorizationError';
    }
}

class RateLimitError extends AppError {
    constructor(retryAfter = 60) {
        super('Rate limit exceeded', 'RATE_LIMITED', 429, { retryAfter });
        this.name = 'RateLimitError';
    }
}

class ConflictError extends AppError {
    constructor(message, details = null) {
        super(message, 'CONFLICT', 409, details);
        this.name = 'ConflictError';
    }
}

// Express error handler middleware
function errorHandler(err, req, res, next) {
    const requestId = req.id || req.headers['x-request-id'];

    // Log the error
    console.error({
        type: 'error',
        requestId,
        error: err.errorCode || 'INTERNAL_ERROR',
        message: err.message,
        stack: process.env.NODE_ENV === 'development' ? err.stack : undefined,
        details: err.details,
    });

    // Handle known errors
    if (err instanceof AppError) {
        return res.status(err.statusCode).json({
            error: err.errorCode,
            message: err.message,
            details: err.details,
            requestId,
        });
    }

    // Handle unknown errors
    res.status(500).json({
        error: 'INTERNAL_ERROR',
        message: process.env.NODE_ENV === 'production'
            ? 'An unexpected error occurred'
            : err.message,
        requestId,
    });
}

// Async handler wrapper to catch async errors
function asyncHandler(fn) {
    return (req, res, next) => {
        Promise.resolve(fn(req, res, next)).catch(next);
    };
}

module.exports = {
    AppError,
    NotFoundError,
    ValidationError,
    AuthenticationError,
    AuthorizationError,
    RateLimitError,
    ConflictError,
    errorHandler,
    asyncHandler,
};
